ray-on-gke/kuberay-tpu-webhook/main.go

```diff
@@ -123,25 +132,138 @@ func injectHostnames(hostNames string, workerGroupSpec ray.WorkerGroupSpec, work
 	}
 }
 
+func checkWorkersMatchTopology(workerGroupSpec ray.WorkerGroupSpec) (bool, error) {
+	replicas := workerGroupSpec.Replicas
+	if replicas == nil {
+		klog.Error("workerGroupSpec replicas not set")
+	}
+	numWorkers := int(*replicas)
+	containers := workerGroupSpec.Template.Spec.Containers
+	if containers == nil {
+		klog.Error("Container path not specified")
+	}
+	if containerRequestingTPUs(containers...) {
+		topology := workerGroupSpec.Template.Spec.NodeSelector["cloud.google.com/gke-tpu-topology"]
+		if topology == "" {
+			klog.Error("TPU topology not specified")
+		}
+		hosts, err := getNumTPUHosts(topology)
+		if err != nil {
+			return false, err
+		}
+
+		if hosts != numWorkers {
+			return false, nil
+		}
+	}
+	return true, nil
+}
+
+func isScheduledWithAffinity(workerGroupSpec ray.WorkerGroupSpec) (bool, error) {
+	containers := workerGroupSpec.Template.Spec.Containers
+	if containers == nil {
+		klog.Error("Container path not specified")
```

Comment:
What criteria are you using to determine when to simply log an error and continue execution, versus return the error? I see both throughout this PR so it is unclear. For many of these errors it seems like you'd want the current execution path to halt and return an error, but for the webhook to keep running, so you'd probably want to return the error.

Reply:
I went through and changed most of the `klog.Error` instances to return an informative error instead, I had been using klog where it seemed like the logic wouldn't continue anyway.